pkg/postage/batchstore/reserve_test.go

```diff
@@ -0,0 +1,306 @@
+}
+
+// TestBatchStoreUnreserve is testing the correct behaviour of the reserve.
+// the following assumptions are tested on each modification of the batches (top up, depth increase, price change)
+// - reserve exceeds capacity
+// - value-consistency of unreserved POs
+func TestBatchStoreUnreserveEvents(t *testing.T) {
+	// temporarily reset reserve Capacity
+	defer func(i int64) {
+		batchstore.Capacity = i
+	}(batchstore.Capacity)
+	batchstore.Capacity = batchstore.Exp2(16)
+
+	bStore, unreserved := setupBatchStore(t)
+	batches := make(map[string]*postage.Batch)
+
+	t.Run("new batches only", func(t *testing.T) {
+		// iterate starting from batchstore.DefaultDepth to maxPO
+		_, depth := batchstore.GetReserve(bStore)
+		for step := 0; depth < swarm.MaxPO; step++ {
+			cs, err := nextChainState(bStore)
+			if err != nil {
+				t.Fatal(err)
+			}
+			var b *postage.Batch
+			if b, err = createBatch(bStore, cs, depth); err != nil {
+				t.Fatal(err)
+			}
+			batches[string(b.ID)] = b
+			if depth, err = checkReserve(bStore, unreserved); err != nil {
+				t.Fatal(err)
+			}
+		}
+	})
+	t.Run("top up batches", func(t *testing.T) {
+		n := 0
+		for id := range batches {
+			b, err := bStore.Get([]byte(id))
+			if err != nil {
+				if errors.Is(storage.ErrNotFound, err) {
+					continue
+				}
+				t.Fatal(err)
+			}
+			cs, err := nextChainState(bStore)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if err = topUp(bStore, cs, b); err != nil {
+				t.Fatal(err)
+			}
+			if _, err = checkReserve(bStore, unreserved); err != nil {
+				t.Fatal(err)
+			}
+			n++
+			if n > len(batches)/5 {
+				break
+			}
+		}
+	})
+	t.Run("dilute batches", func(t *testing.T) {
+		n := 0
+		for id := range batches {
+			b, err := bStore.Get([]byte(id))
+			if err != nil {
+				if errors.Is(storage.ErrNotFound, err) {
+					continue
+				}
+				t.Fatal(err)
+			}
+			cs, err := nextChainState(bStore)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if err = increaseDepth(bStore, cs, b); err != nil {
+				t.Fatal(err)
+			}
+			if _, err = checkReserve(bStore, unreserved); err != nil {
+				t.Fatal(err)
+			}
+			n++
+			if n > len(batches)/5 {
+				break
+			}
+		}
+	})
+}
+
+func TestBatchStoreUnreserveAll(t *testing.T) {
+	// temporarily reset reserve Capacity
+	defer func(i int64) {
+		batchstore.Capacity = i
+	}(batchstore.Capacity)
+	batchstore.Capacity = batchstore.Exp2(16)
+
+	bStore, unreserved := setupBatchStore(t)
+	var batches [][]byte
+	// iterate starting from batchstore.DefaultDepth to maxPO
+	_, depth := batchstore.GetReserve(bStore)
+	for step := 0; depth < swarm.MaxPO; step++ {
+		cs, err := nextChainState(bStore)
+		if err != nil {
+			t.Fatal(err)
+		}
+		event := rand.Intn(6)
+		//  0:  dilute, 1: topup, 2,3,4,5: create
+		var b *postage.Batch
+		if event < 2 && len(batches) > 10 {
+			for {
+				n := rand.Intn(len(batches))
+				b, err = bStore.Get(batches[n])
+				if err != nil {
+					if errors.Is(storage.ErrNotFound, err) {
```

Comment:
the other way around. `errors.Is(err, storage.ErrNotFound)`. Is this actually hit?

Reply:
oops